File: talea-storytelling/scripts/lib/data_inputs.py
"""Resolve scientific inputs declared in config/data-inputs.json."""

import json
import os
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parents[2]
REGISTRY_PATH = REPO_ROOT / "config" / "data-inputs.json"
ENVIRONMENT_VARIABLE = "TALEA_EXTERNAL_DATA"

_registry = None


def external_data_root() -> Path:
    configured = os.environ.get(ENVIRONMENT_VARIABLE)
    if not configured:
        return (REPO_ROOT / "external").resolve()

    root = Path(configured).expanduser()
    if not root.is_absolute():
        root = REPO_ROOT / root
    return root.resolve()


def load_data_inputs() -> dict:
    global _registry
    if _registry is None:
        with REGISTRY_PATH.open("r", encoding="utf8") as registry_file:
            registry = json.load(registry_file)
        if registry.get("schemaVersion") != 1 or not isinstance(registry.get("inputs"), dict):
            raise RuntimeError(f"Unsupported external input registry: {REGISTRY_PATH}")
        _registry = registry["inputs"]
    return _registry


def resolve_data_input(input_id: str) -> Path:
    entry = load_data_inputs().get(input_id)
    if not isinstance(entry, dict) or not isinstance(entry.get("path"), str):
        raise KeyError(f"Unknown external data input: {input_id}")

    relative_path = Path(entry["path"])
    if relative_path.is_absolute():
        raise RuntimeError(f"External data input path must be relative: {input_id}")

    root = external_data_root()
    resolved = (root / relative_path).resolve()
    try:
        resolved.relative_to(root)
    except ValueError as error:
        raise RuntimeError(f"External data input escapes its root: {input_id}") from error
    return resolved


def require_data_input(input_id: str) -> Path:
    resolved = resolve_data_input(input_id)
    if not resolved.is_file():
        raise FileNotFoundError(
            f"Missing external data input '{input_id}': {resolved}. "
            f"Set {ENVIRONMENT_VARIABLE} or populate the default external directory."
        )
    return resolved

File: talea-storytelling/scripts/lib/data-inputs.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const REGISTRY_PATH = path.join(REPO_ROOT, "config/data-inputs.json");
const ENVIRONMENT_VARIABLE = "TALEA_EXTERNAL_DATA";

let registry;

export function externalDataRoot() {
  const configured = process.env[ENVIRONMENT_VARIABLE];
  if (!configured) return path.join(REPO_ROOT, "external");
  return path.isAbsolute(configured)
    ? path.resolve(configured)
    : path.resolve(REPO_ROOT, configured);
}

export function loadDataInputs() {
  if (!registry) {
    const parsed = JSON.parse(fs.readFileSync(REGISTRY_PATH, "utf8"));
    if (parsed.schemaVersion !== 1 || !parsed.inputs || typeof parsed.inputs !== "object") {
      throw new Error(`Unsupported external input registry: ${REGISTRY_PATH}`);
    }
    registry = parsed.inputs;
  }
  return registry;
}

export function resolveDataInput(inputId) {
  const entry = loadDataInputs()[inputId];
  if (!entry || typeof entry.path !== "string") {
    throw new Error(`Unknown external data input: ${inputId}`);
  }
  if (path.isAbsolute(entry.path)) {
    throw new Error(`External data input path must be relative: ${inputId}`);
  }

  const root = externalDataRoot();
  const resolved = path.resolve(root, entry.path);
  const relative = path.relative(root, resolved);
  if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw new Error(`External data input escapes its root: ${inputId}`);
  }
  return resolved;
}

export function requireDataInput(inputId) {
  const resolved = resolveDataInput(inputId);
  if (!fs.existsSync(resolved) || !fs.statSync(resolved).isFile()) {
    throw new Error(
      `Missing external data input '${inputId}': ${resolved}. ` +
        `Set ${ENVIRONMENT_VARIABLE} or populate the default external directory.`,
    );
  }
  return resolved;
}
